PMax Assets: Add step 2 to the campaign editing page

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implements parts of 📌 [Add step 2 to the campaign management page](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-campaign-step2) in #1787.

- Add step 2 to the campaign editing page and the support of starting from step 2
- Make `AdsCampaign` component support editing a campaign
- Add some editing logic to `AssetGroup` component

It also adds a check to facilitate debugging.

- Add a check for the existence of a context provider to `useAdaptiveFormContext`

### Screenshots:

https://user-images.githubusercontent.com/17420811/205005011-a6c2db43-9da9-4829-bada-a9c739e22755.mp4

### Detailed test instructions:

1. Edit a campaign via the edit button on its most right side.
   - When entering the editing page, it should start with the first step.
   - Check if the stepper works correctly.
   - Update campaigns individually using "Save changes" and "Skip adding assets" buttons.
1. Go back to the Dashboard page.
1. Edit a campaign via the edit button on its "Assets" column.
   - When entering the editing page, it should start with the second step.
   - Update campaigns individually using "Save changes" and "Skip adding assets" buttons.

### Changelog entry

eason9487 committed:
Add a new parameter `initialStep` to getEditCampaignUrl function

eason9487 committed:
Add a check for the existence of a context provider to useAdaptiveFormContext

eason9487 committed:
Make AdsCampaign component support editing a campaign

eason9487 committed:
Add some editing logics to AssetGroup component

eason9487 committed:
Add step 2 to the campaign editing page

eason9487 committed:
Add the asset column to programs table to demonstrate directly going to the assets editing step

puntope commented on js/src/components/paid-ads/ads-campaign.js:
```diff
@@ -45,15 +60,24 @@ export default function AdsCampaign( { onContinue } ) {
 			/>
-			<CreateCampaignFormContent formProps={ formContext } />
+			{ isCreation ? (
+				<CreateCampaignFormContent formProps={ formContext } />
+			) : (
+				<EditPaidAdsCampaignFormContent
+					formProps={ formContext }
+					allowMultiple={ campaign.allowMultiple }
+				/>
```
💅  This is what I meant in our previous discussion in https://github.com/woocommerce/google-listings-and-ads/pull/1794#discussion_r1033216007 here is clearer the divergence in naming:

`CreateCampaignFormContent ` and `EditPaidAdsCampaignFormContent` seems not consistent in naming.

I'd suggest this 2 options:

1. `CreateCampaignFormContent` & `EditCampaignFormContent`
2. `CreatePaidAdsCampaignFormContent` & `EditPaidAdsCampaignFormContent`

eason9487 replied:
Sorry, I misunderstood you were asking to rename the `AdsCampaign` to `AdsCampaignForm` or `AdsCampaignFormContent` in https://github.com/woocommerce/google-listings-and-ads/pull/1794#discussion_r1033216007.

Since `CreateCampaignFormContent` and `EditPaidAdsCampaignFormContent` were not implemented in this PR or this project, they are a little out of scope. But it may be a good opportunity to handle some of the name ambiguities and code duplication further.

I opened a separate PR #1803 to:

- Handle inconsistent component names
- Eliminate the ambiguous component hierarchy mentioned in [my previous comment about the *FormContent names](https://github.com/woocommerce/google-listings-and-ads/pull/1794#discussion_r1035576050)

puntope commented on js/src/components/paid-ads/asset-group/asset-group.js:
```diff
@@ -38,22 +48,27 @@ export default function AssetGroup() {
-				>
-					{ __( 'Skip adding assets', 'google-listings-and-ads' ) }
-				</AppButton>
+				{ ( isCreation || isEmptyAssetGroup ) && (
```
Correct me if I'm wrong, but isn't true that in Creation mode the assets will be always empty?

eason9487 replied:
Currently, all campaigns created by GLA always have at least an associated asset group. In GLA, we want to prompt users that the assets could be added more to improve their ads' performance, so there is a logic specific to GLA to identify if an existing asset group is "empty". Ref: pcTzPl-1c9-p2#comment-2001

Therefore, when creating, the asset group is "non-existent", while when editing, the asset group already exists but it needs to distinguish if it's empty. However, the part about determining an empty asset group hasn't been implemented yet, so it's not yet certain whether the asset group is also considered "empty" when creating a new campaign.

puntope commented on js/src/pages/edit-paid-ads-campaign/index.js:
```diff
@@ -55,18 +79,72 @@ const EditPaidAdsCampaign = () => {
+		try {
+			await updateAdsCampaign( campaign.id, { amount } );
+
+			if ( action === ACTION_COMPLETE ) {
```
💅  I didn't mention it last review since I had doubts. But reading it again... I think could be better to rename this to `ACTION_SAVE_ASSETS` to be clearer about what we are doing.

eason9487 replied:
I struggled with this naming while working on this PR. To be precise, the action `ACTION_COMPLETE` refers to saving the campaign and its assets, and the action `ACTION_SKIP_ASSETS` refers to saving the campaign only. So ideally it would be
`ACTION_SUBMIT_CAMPAIGN_AND_ASSETS` and `ACTION_SUBMIT_CAMPAIGN_ONLY`.

But this makes the action naming look inconsistent with the text of the associated buttons, so in the end I chose to match the button texts.

Would you prefer the naming of `ACTION_SUBMIT_CAMPAIGN_AND_ASSETS` and `ACTION_SUBMIT_CAMPAIGN_ONLY`? Or other names?

puntope replied:
Not a strong opinion if it's clear what is doing. So I maybe the two above, or the ones you mentioned also good 👌 

`ACTION_SUBMIT_CAMPAIGN_WITH_ASSETS`
`ACTION_SUBMIT_CAMPAIGN_SKIP_ASSETS`

eason9487 replied:
Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/1799/commits/56af208bb36ac149aa89e19b1c55345b1fac3150.

puntope commented on js/src/dashboard/all-programs-table-card/index.js:
```diff
@@ -130,18 +142,23 @@ const AllProgramsTableCard = ( props ) => {
-								<ProgramToggle key={ el.id } program={ el } />
-							),
+						display: isPaidCampaign && (
+							// TODO: Revisit this temporary demo since the spec of assets column is not yet decided.
```
❓  Could you elaborate what do you mean with this?

eason9487 replied:
The spec of assets column was not yet decided. This column might not be added, or the logic of the column content might still be adjusted. To facilitate the review of this PR, I first made a column for demonstrating entering the asset edit step directly. This `TODO` is used to remind me to come back to handle it, and the reason.

puntope reviewed (commented):
Howdy @eason9487 

Thanks for working on this. Seems like working. I just left a few comments regarding naming to be discussed before approval.

eason9487 commented:
@puntope, thanks for the review.

> I just left a few comments regarding naming to be discussed before approval.

I replied to all the comments. Could you take a new round of code reviews?

eason9487 committed:
Rename the constant names used to identify the action of submit buttons.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1799#discussion_r1038761874

puntope reviewed (approved):
LGTM ✅ 

Thanks for explanations
